opm/simulators/linalg/AmgxPreconditioner.hpp

```diff
@@ -0,0 +1,348 @@
+            amgx_mode = AMGX_mode_dDFI;
+        } else if constexpr (std::is_same_v<matrix_field_type, float> && std::is_same_v<vector_field_type, float>) {
+            amgx_mode = AMGX_mode_dFFI;
+        } else {
+            OPM_THROW(std::runtime_error, "Unsupported combination of matrix and vector types in AmgxPreconditioner");
+        }
+
+        // Create solver and matrix/vector handles with selected mode
+        AMGX_SAFE_CALL(AMGX_solver_create(&solver_, rsrc_, amgx_mode, cfg_));
+        AMGX_SAFE_CALL(AMGX_matrix_create(&A_amgx_, rsrc_, amgx_mode));
+        AMGX_SAFE_CALL(AMGX_vector_create(&x_amgx_, rsrc_, amgx_mode));
+        AMGX_SAFE_CALL(AMGX_vector_create(&b_amgx_, rsrc_, amgx_mode));
+
+        // Setup matrix structure
+        std::vector<int> row_ptrs(N_ + 1);
+        std::vector<int> col_indices(nnz_);
+        setupSparsityPattern(row_ptrs, col_indices);
+
+        // initialize matrix with values
+        const matrix_field_type* values = &(A_[0][0][0][0]);
+        AMGX_SAFE_CALL(AMGX_pin_memory(const_cast<matrix_field_type*>(values), sizeof(matrix_field_type) * nnz_ * block_size * block_size));
+        AMGX_SAFE_CALL(AMGX_matrix_upload_all(A_amgx_, N_, nnz_, block_size, block_size,
+                                             row_ptrs.data(), col_indices.data(),
+                                             values, nullptr));
+        update();
+    }
+
+    /**
+     * @brief Destructor for the AmgxPreconditioner class.
+     *
+     * Cleans up resources allocated by the preconditioner.
+     */
+    ~AmgxPreconditioner()
+    {
+        const matrix_field_type* values = &(A_[0][0][0][0]);
+        AMGX_SAFE_CALL(AMGX_unpin_memory(const_cast<matrix_field_type*>(values)));
+        if (solver_) {
+            AMGX_SAFE_CALL(AMGX_solver_destroy(solver_));
+        }
+        if (x_amgx_) {
+            AMGX_SAFE_CALL(AMGX_vector_destroy(x_amgx_));
+        }
+        if (b_amgx_) {
+            AMGX_SAFE_CALL(AMGX_vector_destroy(b_amgx_));
+        }
+        if (A_amgx_) {
+            AMGX_SAFE_CALL(AMGX_matrix_destroy(A_amgx_));
+        }
+        // Destroying resrouces and config crashes when reinitializing
+        //if (rsrc_) {
+        //    AMGX_SAFE_CALL(AMGX_resources_destroy(rsrc_));
+        //}
+        //if (cfg_) {
+        //    AMGX_SAFE_CALL(AMGX_config_destroy(cfg_));
+        //}
+    }
+
+    /**
+     * @brief Pre-processing step before applying the preconditioner.
+     *
+     * This method is currently a no-op.
+     *
+     * @param v The update vector.
+     * @param d The defect vector.
+     */
+    void pre(X& /*v*/, Y& /*d*/) override {
+    }
+
+    /**
+     * @brief Applies the preconditioner to a vector.
+     *
+     * Performs one AMG cycle to solve the system.
+     * Involves uploading vectors to AMGX, applying the preconditioner, and downloading the result.
```

Comment:
This means this class behaves like a CPU preconditioner outwardly, would it be simple to make it a pure GPU preconditioner instead and use the wrapper to make a CPU preconditioner of it (for combining with the current CPU CPRW etc.)?

Reply:
Yes exactly! I do believe it is possible to make it a pure GPU preconditioner, and maybe reusing the device memory, as both AMGX and gpuISTL uses CuSparse under the hood. This is something I don't have time to do now but I want to explore further next year. So suggesting that we merge it in as it is now and revisit it.